README.md

```diff
@@ -1 +1,292 @@
+    ...
+  }
+}
+```
+
+Go ahead and try to run it (`environment-exports.js`) with this one-liner command:
+
+```
+node -e "console.log(require('./environment-exports')().apiKey)"
+```
+
+If you're getting `TypeError: require(...) is not a function`, that's right. You are getting an error because now nothing has been exported. By writing `exports = ...`, we **lost** the original value of `exports`!
+
+Note: We lose the reference, because we overwrite the object `exports`. It's okay to overwrite `module.exports` because in this case `module.exports` is a property of the `module` object.
+
+Let's take another example of code to refactor. The code below works, but what happens if we just try to use `exports` instead of `module.exports`?
+
+With `module.exports`:
+
+```js
+module.exports = {
+  sayHelloInEnglish: function() {
+    return 'Hello'
+  },
+  sayHelloInTatar: function() {
+    return 'Isänmesez'
+  },
+  sayHelloInSpanish: function() {
+    return 'Hola'
+  },
+  sayHelloInJapanese: function() {
+    return 'こんにちは' // Kon'nichiwa
+  }
+}
+```
+
+With just `exports`:
+```js
+exports = {
+  sayHelloInEnglish: function() {
+    return 'Hello'
+  },
+  sayHelloInTatar: function() {
+    return 'Isänmesez'
+  },
+  sayHelloInSpanish: function() {
+    return 'Hola'
+  },
+  sayHelloInJapanese: function() {
+    return 'こんにちは' // Kon'nichiwa
+  }
+}
+```
+
+So what's happening with just `exports =...`? It won't work as expected, because we cannot assign exports to something. We lose the reference.
+
+We need to use `module.exports =...` if we want to assign the entire module, i.e., if we want to have the logic as the result of `require()` and not as the result of `require().NAME` where `NAME` is some arbitrary name (e.g, `configs`, `users`).
+
+Typicall, Node developers have one `module.exports = ...` in a file or many `exports.NAME =...`. When to use `module.exports = ...` vs. `exports.name = ...` or `module.exports.name = ...`? In other words, when do you want to use `var name = require('name')` and when `var nameB = require('nameA').nameB`?
+
+Node developers tend to use the `module.exports = ...` when they have a large class which takes the whole file. They use the `exports.NAME = ...` when they have a set of individual methods or objects. A good example of the former approach is the Express.js library. An example of the latter is a collection of some utilities which are categorically the same, but have distinct functionalities which don't need to communicate with one another.
+
+An example of the latter is `chai` which has `except` property: `var expect = require('chai').expect`.
+
+You've learned the most common patterns and their differences. From them you can create more advanced implementation of modules which export constructor (psedu-classical inheritance) a higher-order function (function which returns a function), or monkey patch or use globals. Follow the links in the resources to learn more about them.
```

Comment:
it's very well defined term 
http://javascript.info/tutorial/pseudo-classical-pattern